DotVVM.Diagnostics.StatusPage/DotHtmlFilesRuntimePrecompiler.cs

```diff
@@ -0,0 +1,180 @@
+        {
+            this.serviceProvider = serviceProvider;
+            this.dotvvmConfiguration = config;
+            this.statusPageOptions = statusPageOptions;
+        }
+        private readonly Type DotvvmPresenterType = typeof(DotvvmPresenter);
+        private bool IsDotvvmPresenter(RouteBase r)
+        {
+            var presenter = r.GetPresenter(serviceProvider);
+            return presenter.GetType().IsAssignableFrom(DotvvmPresenterType);
+        }
+        public void ResolveAllFiles()
+        {
+            if (Routes != null) return;
+            lock (_locker)
+            {
+                if (Routes != null) return;
+
+                routes = new ReadOnlyCollection<DotHtmlFileInfo>(dotvvmConfiguration.RouteTable.Select(r => new DotHtmlFileInfo()
+                {
+                    VirtualPath = r.VirtualPath,
+                    Url = r.Url,
+                    HasParameters = r.ParameterNames.Any(),
+                    DefaultValues = r.DefaultValues.Select(s => s.Key + ":" + s.Value?.ToString()).ToList(),
+                    RouteName = r.RouteName,
+                    Status = (string.IsNullOrWhiteSpace(r.VirtualPath) || !IsDotvvmPresenter(r))
+                        ? CompilationState.NonCompilable
+                        : CompilationState.None
+                }).ToList());
+
+                controls = new ReadOnlyCollection<DotHtmlFileInfo>(
+                    dotvvmConfiguration.Markup.Controls.Where(s => !string.IsNullOrWhiteSpace(s.Src))
+                    .Select(s => new DotHtmlFileInfo()
+                    {
+                        TagName = s.TagName,
+                        VirtualPath = s.Src,
+                        Namespace = s.Namespace,
+                        Assembly = s.Assembly,
+                        TagPrefix = s.TagPrefix
+                    }).ToList());
+            }
+        }
+
+        internal void BuildView(DotHtmlFileInfo file, ConcurrentBag<DotHtmlFileInfo> tempList)
+        {
+            Debug.WriteLine($"Precompiling path: {file?.VirtualPath}");
+            if (file.Status != CompilationState.NonCompilable)
+            {
+                try
+                {
+                    var controlFactory = serviceProvider.GetRequiredService<IControlBuilderFactory>();
+
+                    var pageBuilder = controlFactory.GetControlBuilder(file.VirtualPath);
+
+                    var compiledControl = pageBuilder.builder.Value.BuildControl(controlFactory, serviceProvider);
+
+                    if (compiledControl is DotvvmView view && view.Directives.TryGetValue(
+                            ParserConstants.MasterPageDirective,
+                            out var masterPage))
+                    {
+                        if (MasterPages.All(s => s.VirtualPath != masterPage) &&
+                            tempList.All(s => s.VirtualPath != masterPage))
```

Comment:
The logic with master pages seems to be cluttered. Could we rather a return list of master pages of that view and care about what do we do about it in one place?

Reply:
Also, due to races this condition does not really prevent duplicates in the `tempList`